optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/DroolsUniAccumulateFunctionAdapter.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.kie.api.runtime.rule.AccumulateFunction;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+import org.optaplanner.core.impl.score.stream.drools.uni.UniConstraintCollectorAdapter;
+
+/**
+ * Drools {@link AccumulateFunction} that calls {@link UniConstraintCollector} underneath.
+ * @param <A> input to accumulate
+ * @param <ResultContainer_> implementation detail
+ * @param <NewA> result of accumulation
+ */
+public final class DroolsUniAccumulateFunctionAdapter<A, ResultContainer_ extends Serializable, NewA>
+        implements AccumulateFunction<ResultContainer_> {
+
+    private final Supplier<ResultContainer_> supplier;
+    private final BiFunction<ResultContainer_, A, Runnable> accumulator;
+    private final Function<ResultContainer_, NewA> finisher;
+    private final Map<Object, Undo> undoMap = new HashMap<>(0);
+
+    public DroolsUniAccumulateFunctionAdapter(UniConstraintCollectorAdapter<A, ResultContainer_, NewA> collector) {
+        this.supplier = collector.get().supplier();
+        this.accumulator = collector.get().accumulator();
+        this.finisher = collector.get().finisher();
+    }
+
+    @Override
+    public ResultContainer_ createContext() {
+        return supplier.get();
+    }
+
+    @Override
+    public void init(ResultContainer_ context) {
+        undoMap.clear();
+    }
+
+    @Override
+    public void accumulate(ResultContainer_ context, Object value) {
+        undoMap.compute(value, (k, v) -> {
+           if (v == null) {
+               Runnable runnable = accumulator.apply(context, (A) value);
+               return new Undo(runnable);
+           } else {
+               v.incrementUseCount();
```

Comment:
Can you do me a favor and proof through debugging that this increment sometimes goes above 1 for some of the examples?

I find it strange that drools would do accumulate(context, Ann) twice without calling reverse(context, Ann) in between.

Reply:
I could go do that, and it may very well turn out that Drools never does that. But this code is still safer, and there's really no extra cost for it. (The map operations would have to be there anyway.) I'm inclined to keep it as is, WDYT?